outlines/base.py

```diff
@@ -0,0 +1,204 @@
+        # Convert args and kwargs to arrays
+        args = [np.array(arg) for arg in args]
+        kwargs = {key: np.array(value) for key, value in kwargs.items()}
+
+        # Broadcast args and kwargs
+        broadcast_shape = np.broadcast(*args, *list(kwargs.values())).shape
+        args = [np.broadcast_to(arg, broadcast_shape) for arg in args]
+        kwargs = {
+            key: np.broadcast_to(value, broadcast_shape)
+            for key, value in kwargs.items()
+        }
+
+        # Flatten args and kwargs
+        flat_args = [np.ravel(arg) for arg in args]
+        flat_kwargs = {key: np.ravel(value) for key, value in kwargs.items()}
+
+        if self.is_coroutine_fn:
+            loop = asyncio.new_event_loop()
+            try:
+                outputs = loop.run_until_complete(
+                    self.vectorize_call_coroutine(
+                        broadcast_shape, flat_args, flat_kwargs
+                    )
+                )
+            finally:
+                loop.close()
+        else:
+            outputs = self.vectorize_call(broadcast_shape, flat_args, flat_kwargs)
+
+        if isinstance(outputs[0], tuple):
+            outputs = tuple(
+                [
+                    np.asarray(x).reshape(broadcast_shape).squeeze()
+                    for x in zip(*outputs)
+                ]
+            )
+        else:
+            outputs = np.asarray(outputs).reshape(broadcast_shape).squeeze()
+
+        return outputs
+
+    def call_with_signature(self, *args, **kwargs):
+        input_core_dims, output_core_dims = self._in_and_out_core_dimensions
+
+        # TODO: Support kwargs
+        if len(kwargs) != 0:
+            raise ValueError(
+                "arguments to a vectorized function must be passed as positional arguments"
+            )
+
+        if len(args) != len(input_core_dims):
+            raise TypeError(
+                "wrong number of positional arguments: "
+                "expected %r, got %r" % (len(input_core_dims), len(args))
+            )
+
+        args = tuple(np.asarray(arg) for arg in args)
+
+        broadcast_shape, dim_sizes = _parse_input_dimensions(args, input_core_dims)
+        input_shapes = _calculate_shapes(broadcast_shape, dim_sizes, input_core_dims)
+        args = [
+            np.broadcast_to(arg, shape, subok=True)
+            for arg, shape in zip(args, input_shapes)
+        ]
+
+        outputs = None
+        nout = len(output_core_dims)
+
+        for index in np.ndindex(*broadcast_shape):
+            results = self.func(*(arg[index] for arg in args))
+            n_results = len(results) if isinstance(results, tuple) else 1
+
+            if n_results == 1:
+                results = (results,)
+
+            if nout != n_results:
+                raise ValueError(
+                    f"wrong number of outputs from the function, expected {nout}, got {n_results}"
+                )
+
+            if outputs is None:
```

Comment:
Perform these operations outside of the loop so we can run coroutine concurrently